### PR TITLE
Support for custom version incrementing rules

### PR DESCRIPTION
**Pull request by vbuell**
For #79

### CODE REVIEW

**Comment by adamdubiel** [comment]
Thanks, i will take a look at it today or tomorrow :)

**Comment by adamdubiel** [comment]
The code looks nice, i will run through it in Idea to see how things connect.
@szpak - we talked about something similar at Geecon, can you take a look at this pull request and see if it would solve your problem?

File: docs/configuration/version.rst [hunk 1]
```diff
@@ -119,6 +119,35 @@ version::
...
+.. _version-incrementing:
+
+Incrementing
+------------
+
+Incrementing phase does incrementing the version in accordance with *version incrementer* rule. By default version patch
+(least significant) number is incremented. There are predefined rules:
+
+- *incrementPatch* - increment patch number
+- *incrementMinor* - increment minor (middle) number
```
Comment (szpak) [inline docs/configuration/version.rst]: There is an inconsistency between `incrementMinor` in a documentation and `incrementMinorVersion` in code. The same for `incrementPatch`.

Comment (vbuell) [inline docs/configuration/version.rst]: Oops. Good catch.
Fixed.

**Comment by szpak** [comment]
@adamdubiel It's not a problem. It's a feature :).

It took a while to make it work, as my project was on a release tag and due to #17 I had no effect using the new parameter :). In the end I was able to do simple custom transformation as expected and after tweaking my code to do a Git lookup to parse the latest commit message in the `versionIncrementingRule` closure I should be able to set a new version using Git commit message.

Btw, guys, have you managed how to install plugin to a local repo with the new `plugin-publish-plugin`?

**Comment by szpak** [comment]
Btw2, really nice work @vbuell !

**Comment by vbuell** [comment]
Adam, all comments are addressed. Please review.

**Comment by adamdubiel** [comment]
Sorry, i suffer from constant lack of time.. there is just one thing i would like to change. Imagine these dynamic version closures might require different arguments (like `incrementMinorIfNotOnRelease` needs `releaseBranchPattern`). I would not want to add a lot of these optional arguments. So maybe they should be configured:

* using map next to definition (e.x. `versionIncrementingRule 'incrementPatch' [ releaseBranchPattern: 'whatever*' ]`
* using parameters in closure, i.e. `versionIncrementingArguments {  releaseBranchPattern: 'whatever*' } ` , arguments in closure should be dynamic (like in map)

Personally i think the first option would be the best, this would also stay consistent with the way pre/post hooks are configured. Other than that - looks great :)

**Comment by vbuell** [comment]
Done

**Comment by adamdubiel** [comment]
Hm, i can still see this Pattern in ScmVersion and no Map for arguments of incrementer - can you tell me how do you understand my last comment? I like the introduction of Context object, it will be helpful in future.

**Comment by vbuell** [comment]
Oh. I did not read it carefully. Sorry. Will redo.
In your example you suggest to move `releaseBranchPattern` from top-level. But if (when) you pull #78, release branch name will be duplicated in two places. I'd avoid that

**Comment by adamdubiel** [comment]
Yeah, i was wondering about it as well.. can you make the change to be able to pass arguments *in general* and we can discuss what to do with this `releaseBranchPattern`.

**Comment by vbuell** [comment]
It's not yagni :D

**Comment by adamdubiel** [comment]
Oh, i already have use case for this :)

**Comment by vbuell** [comment]
You intrigued me. :) Done

**Comment by adamdubiel** [comment]
Thanks - i hope the build passes as well :) I'm going to spend some time with axion during the weekend, merging stuff, adding some small bugfixes and i would like to have 1.3.0 released on Sunday.

**Comment by vbuell** [comment]
Forgot to say that I did not document this configuration map since there in no any predefined incrementers that use it yet. Let's wait till the first consumer of this map appeared

**Comment by adamdubiel** [comment]
Exactly, that's fine.

**Comment by vbuell** [comment]
Thanks for merging!

**Pull request merged**